ion/textwriter.go

```diff
@@ -277,47 +288,105 @@ func (w *textWriter) writeValue(api string, val string) error {
 // annotations (if any).
 func (w *textWriter) beginValue(api string) error {
 	if w.needsSeparator {
-		var sep byte
-		switch w.ctx.peek() {
-		case ctxInStruct, ctxInList:
-			sep = ','
-		case ctxInSexp:
-			sep = ' '
-		default:
-			sep = '\n'
+		if err := w.writeSeparator(); err != nil {
+			return err
 		}
+	}
 
-		if err := writeRawChar(sep, w.out); err != nil {
+	if w.emptyContainer {
+		if w.pretty() {
+			if err := writeRawChar('\n', w.out); err != nil {
+				return err
+			}
+		}
+	}
+
+	if w.pretty() {
+		if err := w.writeIndent(); err != nil {
 			return err
 		}
 	}
 
 	if w.inStruct() {
-		if w.fieldName == "" {
-			return &UsageError{api, "field name not set"}
+		if err := w.writeFieldName(api); err != nil {
+			return err
 		}
-		name := w.fieldName
-		w.fieldName = ""
+	}
 
-		if err := writeSymbol(name, w.out); err != nil {
+	if len(w.annotations) > 0 {
+		if err := w.writeAnnotations(); err != nil {
 			return err
 		}
-		if err := writeRawChar(':', w.out); err != nil {
-			return err
+	}
+
+	return nil
+}
+
+// writeSeparator writes out the character or characters that separate values.
+func (w *textWriter) writeSeparator() error {
+	var sep string
+
+	switch w.ctx.peek() {
+	case ctxInStruct, ctxInList:
+		// In a struct or a list, values are separated by commas.
+		if w.pretty() {
+			sep = ",\n"
+		} else {
+			sep = ","
+		}
+
+	case ctxInSexp:
+		// In an sexp, values are separated by whitespace.
+		if w.pretty() {
+			sep = "\n"
+		} else {
+			sep = " "
 		}
+
+	default:
+		// At the top level, values are separated by newlines.
+		sep = "\n"
 	}
 
-	if len(w.annotations) > 0 {
-		as := w.annotations
-		w.annotations = nil
+	return writeRawString(sep, w.out)
+}
 
-		for _, a := range as {
-			if err := writeSymbol(a, w.out); err != nil {
-				return err
-			}
-			if err := writeRawString("::", w.out); err != nil {
-				return err
-			}
+// writeFieldName writes a field name inside a struct.
+func (w *textWriter) writeFieldName(api string) error {
+	if w.fieldName == "" {
+		return &UsageError{api, "field name not set"}
+	}
+	name := w.fieldName
+	w.fieldName = ""
+
+	if err := writeSymbol(name, w.out); err != nil {
+		return err
+	}
+
+	sep := ":"
+	if w.pretty() {
+		sep = " : "
+	}
+
+	return writeRawString(sep, w.out)
+}
+
+// writeAnnotations writes out the annotations for a value.
+func (w *textWriter) writeAnnotations() error {
+	as := w.annotations
+	w.annotations = nil
+
+	sep := "::"
+	if w.pretty() {
+		sep = " :: "
+	}
```

Comment:
We do not have a spec for pretty printing. The libs are essentially free to generate something that is easy for humans to read, id valid Ion Text and round trips to the same values. Having said that (and it may be that I am too used to what other Ion libs do) for the annotation's separator typically there are no spaces around `::`. It will make the code also simpler :D

Reply:
Hah, this is how I used to "pretty print" Ion by hand back in the day. Happy to follow what the other implementations do, especially if it means I get to delete code. :)